scapy/asn1/asn1.py

```diff
@@ -73,57 +99,74 @@ class ASN1_BadTag_Decoding_Error(ASN1_Decoding_Error):
 
 class ASN1Codec(EnumElement):
     def register_stem(cls, stem):
+        # type: (Type[BERcodec_Object[Any]]) -> None
         cls._stem = stem
 
     def dec(cls, s, context=None):
-        return cls._stem.dec(s, context=context)
+        # type: (bytes, Optional[Type[ASN1_Class]]) -> ASN1_Object[Any]
+        return cls._stem.dec(s, context=context)  # type: ignore
 
     def safedec(cls, s, context=None):
-        return cls._stem.safedec(s, context=context)
+        # type: (bytes, Optional[Type[ASN1_Class]]) -> ASN1_Object[Any]
+        return cls._stem.safedec(s, context=context)  # type: ignore
 
     def get_stem(cls):
-        return cls.stem
+        # type: () -> type
+        return cls._stem
 
 
 class ASN1_Codecs_metaclass(Enum_metaclass):
     element_class = ASN1Codec
 
 
-class ASN1_Codecs(six.with_metaclass(ASN1_Codecs_metaclass)):
-    BER = 1
-    DER = 2
-    PER = 3
-    CER = 4
-    LWER = 5
-    BACnet = 6
-    OER = 7
-    SER = 8
-    XER = 9
+@six.add_metaclass(ASN1_Codecs_metaclass)
+class ASN1_Codecs:
+    BER = cast(ASN1Codec, 1)
+    DER = cast(ASN1Codec, 2)
+    PER = cast(ASN1Codec, 3)
+    CER = cast(ASN1Codec, 4)
+    LWER = cast(ASN1Codec, 5)
+    BACnet = cast(ASN1Codec, 6)
+    OER = cast(ASN1Codec, 7)
+    SER = cast(ASN1Codec, 8)
+    XER = cast(ASN1Codec, 9)
 
 
 class ASN1Tag(EnumElement):
-    def __init__(self, key, value, context=None, codec=None):
+    def __init__(self,
+                 key,  # type: str
+                 value,  # type: int
+                 context=None,  # type: Optional[Type[ASN1_Class]]
+                 codec=None  # type: Optional[Dict[ASN1Codec, Type[BERcodec_Object[Any]]]]  # noqa: E501
+                 ):
+        # type: (...) -> None
         EnumElement.__init__(self, key, value)
-        self._context = context
+        # populated by the metaclass
+        self.context = context  # type: Type[ASN1_Class]  # type: ignore
         if codec is None:
             codec = {}
         self._codec = codec
 
     def clone(self):  # not a real deep copy. self.codec is shared
-        return self.__class__(self._key, self._value, self._context, self._codec)  # noqa: E501
+        # type: () -> Any
```

Comment:
Why cannot you have this here?
```suggestion
        # type: () -> ASN1Tag
```

Reply:
Good catch